File: src/main/julia/marblemania.jl
mutable struct Marble
    value::Int
    next::Marble
    prev::Marble

    function Marble(v)
        m = new(v)
        m.prev = m.next = m
    end
end

function appendMarble(m::Marble, newm::Marble)
    newm.prev = m
    newm.next = m.next
    m.next.prev = newm
    m.next = newm
end


function removeMarble(m::Marble)
    m.prev.next = m.next
    m.next.prev = m.prev
end

function play(nplayers, lastmarble)
    score = [0 for p = 1:nplayers]
    current = Marble(0)

    for val = 1:lastmarble
        player = 1+ val % nplayers

        if val % 23 == 0
            r = current.prev.prev.prev.prev.prev.prev.prev
            removeMarble(r)
            current = r.next
            score[player] += val + r.value
        else
            m = Marble(val)
            appendMarble(current.next, m)
            current = m
        end
    end

    max(score...)
end

play(10, 1618)

@show play(427, 70723)

@show play(427, 7072300)

File: src/main/go/marblemania.go
package main

// Marble is the structure for our game
type Marble struct {
	value int
	next  *Marble
	prev  *Marble
}

// NewMarble is the factory function for Marble
func NewMarble(value int) *Marble {
	m := &Marble{value, nil, nil}
	return m
}

func (m *Marble) insertBetween(a, b *Marble) {
	m.prev = a
	m.next = b
	a.next = m
	b.prev = m
}

func (m *Marble) remove() {
	m.prev.next = m.next
	m.next.prev = m.prev
}

func play(players, marbles int) int {
	score := make([]int, players)

	current := NewMarble(0)
	current.next = current
	current.prev = current

	for i := 1; i < marbles; i++ {
		if i%23 == 0 {
			m := current.prev.prev.prev.prev.prev.prev.prev
			m.remove()
			score[i%players] += i
			score[i%players] += m.value
			current = m.next

		} else {
			m := NewMarble(i)
			m.insertBetween(current.next, current.next.next)
			current = m
		}

	}
	return max(score)
}

func max(score []int) int {
	m := 0
	for _, v := range score {
		if v > m {
			m = v
		}
	}
	return m
}

func main() {
	println(play(10, 1618))
	println(play(427, 70723))
}
